File: xy10938/src/services/memberService.js
const { runQuery, getOne, getAll } = require('../db');

function generateMemberNo() {
  const lastMember = getOne(`SELECT member_no FROM members ORDER BY id DESC LIMIT 1`);
  let seq = 1;
  if (lastMember) {
    const match = lastMember.member_no.match(/M(\d{6})/);
    if (match) {
      seq = parseInt(match[1]) + 1;
    }
  }
  return `M${String(seq).padStart(6, '0')}`;
}

function createMember(data) {
  const { name, phone, level = '普通会员', balance = 0 } = data;

  if (!name || !phone) {
    throw { status: 400, message: '姓名和手机号不能为空', conclusion: '必填参数缺失' };
  }

  const existing = getOne(`SELECT id FROM members WHERE phone = ?`, [phone]);
  if (existing) {
    throw { status: 400, message: '该手机号已注册', conclusion: '拦截重复注册请求' };
  }

  const memberNo = generateMemberNo();
  const result = runQuery(
    `INSERT INTO members (member_no, name, phone, level, balance) VALUES (?, ?, ?, ?, ?)`,
    [memberNo, name, phone, level, balance]
  );

  return getOne(`SELECT * FROM members WHERE id = ?`, [result.lastInsertRowid]);
}

function getMemberList() {
  return getAll(`SELECT * FROM members ORDER BY created_at DESC`);
}

function getMemberById(id) {
  return getOne(`SELECT * FROM members WHERE id = ?`, [id]);
}

function getMemberByPhone(phone) {
  return getOne(`SELECT * FROM members WHERE phone = ?`, [phone]);
}

function updateMember(id, data) {
  const member = getMemberById(id);
  if (!member) {
    throw { status: 404, message: '会员不存在', conclusion: '找不到对应会员记录' };
  }

  const allowedFields = ['name', 'phone', 'level', 'balance'];
  const updates = [];
  const params = [];

  for (const field of allowedFields) {
    if (data[field] !== undefined) {
      updates.push(`${field} = ?`);
      params.push(data[field]);
    }
  }

  if (updates.length === 0) {
    throw { status: 400, message: '没有有效更新字段', conclusion: '请求参数无效' };
  }

  params.push(id);
  runQuery(`UPDATE members SET ${updates.join(', ')} WHERE id = ?`, params);

  return getMemberById(id);
}

module.exports = {
  createMember,
  getMemberList,
  getMemberById,
  getMemberByPhone,
  updateMember
};

File: xy10938/services/member_service.go
package services

import (
	"car-wash-queue-api/database"
	"car-wash-queue-api/models"
	"database/sql"
	"fmt"
)

func generateMemberNo() (string, error) {
	var lastMemberNo string
	err := database.DB.QueryRow(`
		SELECT member_no FROM members ORDER BY id DESC LIMIT 1
	`).Scan(&lastMemberNo)
	
	if err == sql.ErrNoRows {
		return "M000001", nil
	}
	if err != nil {
		return "", err
	}

	var seq int
	fmt.Sscanf(lastMemberNo, "M%d", &seq)
	return fmt.Sprintf("M%06d", seq+1), nil
}

func CreateMember(name, phone string, params ...interface{}) (*models.Member, error) {
	if name == "" || phone == "" {
		return nil, fmt.Errorf("姓名和手机号不能为空")
	}

	level := "普通会员"
	balance := 0.0

	if len(params) >= 1 {
		if l, ok := params[0].(string); ok {
			level = l
		}
	}
	if len(params) >= 2 {
		if b, ok := params[1].(float64); ok {
			balance = b
		}
	}

	var existingID int
	err := database.DB.QueryRow(`SELECT id FROM members WHERE phone = ?`, phone).Scan(&existingID)
	if err != sql.ErrNoRows {
		if err == nil {
			return nil, fmt.Errorf("该手机号已注册")
		}
		return nil, err
	}

	memberNo, err := generateMemberNo()
	if err != nil {
		return nil, err
	}

	result, err := database.DB.Exec(`
		INSERT INTO members (member_no, name, phone, level, balance)
		VALUES (?, ?, ?, ?, ?)
	`, memberNo, name, phone, level, balance)
	if err != nil {
		return nil, err
	}

	id, err := result.LastInsertId()
	if err != nil {
		return nil, err
	}

	return GetMemberByID(int(id))
}

func GetMemberByID(id int) (*models.Member, error) {
	var member models.Member
	err := database.DB.QueryRow(`
		SELECT id, member_no, name, phone, level, balance, created_at, updated_at
		FROM members WHERE id = ?
	`, id).Scan(&member.ID, &member.MemberNo, &member.Name, &member.Phone,
		&member.Level, &member.Balance, &member.CreatedAt, &member.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &member, nil
}

func GetMemberByPhone(phone string) (*models.Member, error) {
	var member models.Member
	err := database.DB.QueryRow(`
		SELECT id, member_no, name, phone, level, balance, created_at, updated_at
		FROM members WHERE phone = ?
	`, phone).Scan(&member.ID, &member.MemberNo, &member.Name, &member.Phone,
		&member.Level, &member.Balance, &member.CreatedAt, &member.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &member, nil
}

func GetMemberList() ([]models.Member, error) {
	rows, err := database.DB.Query(`
		SELECT id, member_no, name, phone, level, balance, created_at, updated_at
		FROM members ORDER BY created_at DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var members []models.Member
	for rows.Next() {
		var m models.Member
		err := rows.Scan(&m.ID, &m.MemberNo, &m.Name, &m.Phone,
			&m.Level, &m.Balance, &m.CreatedAt, &m.UpdatedAt)
		if err != nil {
			return nil, err
		}
		members = append(members, m)
	}
	return members, nil
}

func UpdateMember(id int, data map[string]interface{}) (*models.Member, error) {
	_, err := GetMemberByID(id)
	if err != nil {
		return nil, fmt.Errorf("会员不存在")
	}

	updates := ""
	params := []interface{}{}
	first := true

	allowedFields := []string{"name", "phone", "level", "balance"}
	for _, field := range allowedFields {
		if val, ok := data[field]; ok {
			if !first {
				updates += ", "
			}
			updates += fmt.Sprintf("%s = ?", field)
			params = append(params, val)
			first = false
		}
	}

	if updates == "" {
		return nil, fmt.Errorf("没有有效更新字段")
	}

	params = append(params, id)
	_, err = database.DB.Exec(`UPDATE members SET `+updates+` WHERE id = ?`, params...)
	if err != nil {
		return nil, err
	}

	return GetMemberByID(id)
}
